ovnkube-node: Existing management port check

DmytroLinkin opened the pull request:
When ovnkube-node configures management port interface it may fail if host already have it configured but it is different interface. To handle that do the following:
- verify that port exists on the integrational bridge and it represents provided management port netdevice or OVS internal port, if no netdevice were provided;
- verify that no interface exists with the name of the management port;

If any of the checks had failed then need to remove port from the bridge and if it wasn't OVS internal port - flush IP addresses, routes, tables and rename the interface.

Because ovnkube doesn't have unload sequence, port check done during new management port creation.

Signed-off-by: Dmytro Linkin <dlinkin@nvidia.com>

DmytroLinkin commented:
/cc @zshi-redhat @wizhaoredhat

zshi-redhat commented on go-controller/pkg/node/management-port_linux.go:
```diff
@@ -280,6 +293,108 @@ func createPlatformManagementPort(interfaceName string, localSubnets []*net.IPNe
+}
+
+func unconfigureMgmtRepresentorPort(mgmtPortName string) error {
+	_, stderr, err := util.RunOVSVsctl("del-port", "br-int", mgmtPortName)
```
Add `--if-exists` in the ovsctl run?

zshi-redhat replied:
Agree, it is not strictly needed here.

naserdean replied:
@DmytroLinkin please add the argument `--if-exists`.
This would be good just as a safety measure if someone outside of the scope of ovn-kubernetes makes any operations to that port.

zshi-redhat commented on go-controller/pkg/node/management-port_linux.go:
```diff
@@ -280,6 +293,108 @@ func createPlatformManagementPort(interfaceName string, localSubnets []*net.IPNe
+	}
+
+	// rename to PortName + "-ddmmyyHHMMSS"
+	if err := util.GetNetLinkOps().LinkSetName(link, mgmtPortName+time.Now().Format("-010206150405")); err != nil {
```
Is the actual netdev or representor dev renamed here? or does it become orphaned device?

DmytroLinkin replied:
Only representor can be found on the bridge. So it is renamed to avoid conflict.
Didn't get Q. about orphaned device.

zshi-redhat replied:
I was wondering whether the renamed devices become useless. 
If I read it correctly, both devices are renamed. The representor device is renamed in `unconfigureMgmtRepresentorPort` and netdev device is renamed in `unconfigurePlatformMgmtInterface`.  The renamed device pair can be discovered by sriov operator and reused for workload pods. do you agree?

DmytroLinkin replied:
Can't say about operator, but sriov-device-plugin look only onto device IDs (VFs - PCI, SFs - aux.names), so netdevice will be added into the resource pool. Ovnkube, in turns, will discover representor, whichever name it has.

DmytroLinkin replied:
The only problem, which actually can be, is when 1st pair of devices were used. Then 2nd one -> 1st pair renamed to templated name. Then user switched back to the 1st pair, but ovnkube fails to find devices under original names either mgmt port names.

zshi-redhat replied:
> The only problem, which actually can be, is when 1st pair of devices were used. Then 2nd one -> 1st pair renamed to templated name. Then user switched back to the 1st pair, but ovnkube fails to find devices under original names either mgmt port names.

Does it make sense to rename the device to its original name by using external-ids?

DmytroLinkin replied:
I'd thought about it. Probably yes. Limitation here is that for DPU-host mode it is not possible. Perhaps it'll be a bit confusing to the user, that in full and DPU modes original names are restored, but not DPU-host. Let me check how it will work/look for full and DPU modes.

DmytroLinkin commented on go-controller/pkg/node/management-port.go:
```diff
@@ -109,6 +113,31 @@ func (mp *managementPort) Create(nodeAnnotator kube.Annotator, waiter *startupWa
+		if !strings.Contains(stderr, "no row \""+types.K8sMgmtIntfName+"\" in table") {
+			return fmt.Errorf("failed to lookup OVS port: %s", stderr)
+		}
+		// Not found on the bridge. But could be that interface with the same name exists
```
> Could this legacy interface be representor port?

Have no idea. But AFAIU from comparing with MgmtNetdev and MgmtRep it is only OVS internal port.

> How do we know only platform mgmt interface needs to be handled here?

What else?

zshi-redhat replied:
> > Could this legacy interface be representor port?
> 
> Have no idea. But AFAIU from comparing with MgmtNetdev and MgmtRep it is only OVS internal port.

From HW full mode to OVS internal port:
1) the mgmt port name was `ovn-k8s-mp0_0` in HW full mode
2) `ovs-vsctl list interface ovn-k8s-mp0` return `no row in the table`
3) it goes to unconfigurePlatformInterface with `ovn-k8s-mp0` which renames the VF netdev to `ovn-k8s-mp0-ddmmyyHHMMSS` and return success
4) this leaves `ovn-k8s-mp0_0` remain in `br-int` bridge

> 
> > How do we know only platform mgmt interface needs to be handled here?
> 
> What else?

The representor port left in step 4, should it be renamed or released?

DmytroLinkin replied:
> > > Could this legacy interface be representor port?
> > 
> > 
> > Have no idea. But AFAIU from comparing with MgmtNetdev and MgmtRep it is only OVS internal port.
> 
> From HW full mode to OVS internal port:
> 
> 1. the mgmt port name was `ovn-k8s-mp0_0` in HW full mode
> 2. `ovs-vsctl list interface ovn-k8s-mp0` return `no row in the table`
> 3. it goes to unconfigurePlatformInterface with `ovn-k8s-mp0` which renames the VF netdev to `ovn-k8s-mp0-ddmmyyHHMMSS` and return success
> 4. this leaves `ovn-k8s-mp0_0` remain in `br-int` bridge
> 

Ok, lets clarify. You are talking about DPU full representor. Not about `legacyMgmtIntfName := util.GetLegacyK8sMgmtIntfName(mp.nodeName)`.

> > > How do we know only platform mgmt interface needs to be handled here?
> > 
> > What else?
> 
> The representor port left in step 4, should it be renamed or released?

Renamed and removed, I suppose, because of flows, You mentioned below.

DmytroLinkin replied:
You know. I had a talk to team mate and in multicluster scenario we actually have two representors on the br-int and it works.

zshi-redhat replied:
> > > > Could this legacy interface be representor port?
> > > 
> > > 
> > > Have no idea. But AFAIU from comparing with MgmtNetdev and MgmtRep it is only OVS internal port.
> > 
> > 
> > From HW full mode to OVS internal port:
> > 
> > 1. the mgmt port name was `ovn-k8s-mp0_0` in HW full mode
> > 2. `ovs-vsctl list interface ovn-k8s-mp0` return `no row in the table`
> > 3. it goes to unconfigurePlatformInterface with `ovn-k8s-mp0` which renames the VF netdev to `ovn-k8s-mp0-ddmmyyHHMMSS` and return success
> > 4. this leaves `ovn-k8s-mp0_0` remain in `br-int` bridge
> 
> Ok, lets clarify. You are talking about DPU full representor. Not about `legacyMgmtIntfName := util.GetLegacyK8sMgmtIntfName(mp.nodeName)`.
> 
yes, HW full offload, not legacy k8s mgmtn interface name.

zshi-redhat replied:
Could you elaborate on the multicluster scenario? Does it address the remaining ovn-k8s-mp0_0 issue from HW full mode to OVS internal port in single cluster?

DmytroLinkin replied:
In Multicluster two instances of ovnkube-node are running: in DPU mode and full mode. To make them not fight over all resources, they are separated, including management ports, so there are two management ports with different names added to br-int. AFAIK, there is no changes to separate flows.

DmytroLinkin commented on go-controller/pkg/node/management-port.go:
```diff
@@ -109,6 +113,31 @@ func (mp *managementPort) Create(nodeAnnotator kube.Annotator, waiter *startupWa
+	// We may reach this point if it is DPU node and previously ovnkube was running in the DPU mode or full mode using
+	// HW backed device, but switched to OVS internal port now.
+	klog.Infof("Found existing representor management port. Removing it")
+	return unconfigureMgmtRepresentorPort(types.K8sMgmtIntfName)
```
Yes, it should be `ovn-k8s-mp0_0`, unless it was OVS internal port `ovn-k8s-mp0`.
Don't know how big is the problem of not deleted representor.

zshi-redhat replied:
I think the consequence is there will be two ovn mgmt ports in br-int bridge: one is ovn-k8s-mp0 with ovs internal port type, the other is ovn-k8s- mp0_0 representor port. OVS flows may not be cleaned up for the representor port which could potentially break traffic that is supposed to traverse through the mgmt port from ovn network.

DmytroLinkin replied:
I also don't know if flows are cleaned. Lets assume no and need to remove the port.

DmytroLinkin commented on go-controller/pkg/node/management-port_linux.go:
```diff
@@ -280,6 +293,108 @@ func createPlatformManagementPort(interfaceName string, localSubnets []*net.IPNe
+	// It is other representor which was used as management port.
+	// Remove it from the bridge and rename.
+	klog.Infof("Found existing representor management port. Removing it")
+	return unconfigureMgmtRepresentorPort(mgmtPortName)
```
What do You mean _doesn't apply to netdev interface_? If OVS port with the same name will be founded on the bridge it will be deleted in both cases.

zshi-redhat replied:
I was wrong,  actually the `unconfigureMgmtRepresentorPort` is only called once when switching from one netdev to another netdev (e.g. vf0 to vf1) in HW full mode. When switching from HW full mode to OVS internal port, it is not called.

wizhaoredhat commented:
I tested this in a local build. Looks good so far. I did not have to manually remove the port ovn-k8s-mp0 OVS port.

DmytroLinkin commented:
/restart_failed

DmytroLinkin commented:
/retest-failed

DmytroLinkin commented:
/retest-failed

wizhaoredhat commented:
Hi Dmytro, with your latest commits (fixups). It seems like there is some issues. I no longer see ovn-k8s-mp0_0 come up. This is the errors I get:
```
ovnkube-node logs:
+ exec /usr/bin/ovnkube --init-node worker-advnetlab48 --nb-address ssl:192.168.111.20:9641,ssl:192.168.111.21:9641,ssl:192.168.111.22:9641 --sb-address ssl:192.168.111.20:9642,ssl:192.168.111.21:9642,ssl:192.168.111.22:9642 --nb-client-privkey /ovn-cert/tls.key --nb-client-cert /ovn-cert/tls.crt --nb-client-cacert /ovn-ca/ca-bundle.crt --nb-cert-common-name ovn --sb-client-privkey /ovn-cert/tls.key --sb-client-cert /ovn-cert/tls.crt --sb-client-cacert /ovn-ca/ca-bundle.crt --sb-cert-common-name ovn --config-file=/run/ovnkube-config/ovnkube.conf --loglevel 4 --inactivity-probe=180000 --gateway-mode shared --gateway-interface br-ex --metrics-bind-address 127.0.0.1:29103 --ovn-metrics-bind-address 127.0.0.1:29105 --metrics-enable-pprof --export-ovs-metrics --disable-snat-multiple-gws --ovnkube-node-mgmt-port-netdev=ens1f0v0
...
I1118 21:01:33.316902   51185 management-port-dpu.go:199] Setup management port dpu host: ens1f0v0
I1118 21:01:33.419318   51185 ovs.go:200] Exec(5): /usr/bin/ovs-vsctl --timeout=15 set Open_vSwitch . external-ids:ovn-orig-mgmt-port-netdev-name=ens1f0v0
I1118 21:01:33.430632   51185 ovs.go:203] Exec(5): stdout: ""
I1118 21:01:33.430650   51185 ovs.go:204] Exec(5): stderr: ""
I1118 21:01:33.584049   51185 ovs.go:200] Exec(6): /usr/bin/ovs-vsctl --timeout=15 list Interface ovn-k8s-mp0_0
I1118 21:01:33.595037   51185 ovs.go:203] Exec(6): stdout: "_uuid               : 4122a940-aecc-430f-bfcd-4fa8af247317\nadmin_state         : up\nbfd                 : {}\nbfd_status          : {}\ncfm_fault           : []\ncfm_fault_status    : []\ncfm_flap_count      : []\ncfm_health          : []\ncfm_mpid            : []\ncfm_remote_mpids    : []\ncfm_remote_opstate  : []\nduplex              : []\nerror               : []\nexternal_ids        : {iface-id=k8s-worker-advnetlab48, ovn-installed=\"true\", ovn-installed-ts=\"1668805274323\", ovn-orig-mgmt-port-rep-name=ens1f0_0}\nifindex             : 20\ningress_policing_burst: 0\ningress_policing_kpkts_burst: 0\ningress_policing_kpkts_rate: 0\ningress_policing_rate: 0\nlacp_current        : []\nlink_resets         : 1\nlink_speed          : []\nlink_state          : up\nlldp                : {}\nmac                 : []\nmac_in_use          : \"f6:a2:e9:2f:10:0a\"\nmtu                 : 1400\nmtu_request         : []\nname                : ovn-k8s-mp0_0\nofport              : 14\nofport_request      : []\noptions             : {}\nother_config        : {}\nstatistics          : {collisions=0, rx_bytes=1162, rx_crc_err=0, rx_dropped=0, rx_errors=0, rx_frame_err=0, rx_missed_errors=0, rx_over_err=0, rx_packets=11, tx_bytes=774, tx_dropped=0, tx_errors=0, tx_packets=7}\nstatus              : {driver_name=mlx5e_rep, driver_version=\"4.18.0-372.31.1.el8_6.mr3440_22\", firmware_version=\"24.35.1012 (MT_0000000540)\"}\ntype                : \"\"\n"
I1118 21:01:33.595088   51185 ovs.go:204] Exec(6): stderr: ""
I1118 21:01:33.595098   51185 ovs.go:200] Exec(7): /usr/bin/ovs-vsctl --timeout=15 get Interface ovn-k8s-mp0_0 type
I1118 21:01:33.605397   51185 ovs.go:203] Exec(7): stdout: "\"\"\n"
I1118 21:01:33.605412   51185 ovs.go:204] Exec(7): stderr: ""
I1118 21:01:33.605419   51185 management-port_linux.go:365] Found existing representor management port. Removing it
I1118 21:01:33.605430   51185 ovs.go:200] Exec(8): /usr/bin/ovs-vsctl --timeout=15 --if-exists get Interface ovn-k8s-mp0_0 external-ids:ovn-orig-mgmt-port-rep-name
I1118 21:01:33.615509   51185 ovs.go:203] Exec(8): stdout: "ens1f0_0\n"
I1118 21:01:33.615525   51185 ovs.go:204] Exec(8): stderr: ""
I1118 21:01:33.615538   51185 ovs.go:200] Exec(9): /usr/bin/ovs-vsctl --timeout=15 --if-exists del-port br-int ovn-k8s-mp0_0
I1118 21:01:33.641359   51185 ovs.go:203] Exec(9): stdout: ""
I1118 21:01:33.641379   51185 ovs.go:204] Exec(9): stderr: ""
I1118 21:01:33.731455   51185 management-port-dpu.go:64] Create representor management port: ovn-k8s-mp0_0
I1118 21:01:33.795935   51185 ovs.go:200] Exec(10): /usr/bin/ovs-vsctl --timeout=15 -- --may-exist add-port br-int ovn-k8s-mp0_0 -- set interface ovn-k8s-mp0_0 external-ids:iface-id=k8s-worker-advnetlab48 external-ids:ovn-orig-mgmt-port-rep-name=ovn-k8s-mp0_0
I1118 21:01:33.812274   51185 ovs.go:203] Exec(10): stdout: ""
I1118 21:01:33.812290   51185 ovs.go:204] Exec(10): stderr: "ovs-vsctl: Error detected while setting up 'ovn-k8s-mp0_0': could not open network device ovn-k8s-mp0_0 (No such device).  See ovs-vswitchd log for details.\novs-vsctl: The default log directory is \"/var/log/openvswitch\".\n"
I1118 21:01:33.812334   51185 gateway_init.go:259] Initializing Gateway Functionality
```
```
In ovs-vswitchd.log:
2022-11-18T21:01:14.307Z|00239|bridge|INFO|bridge br-int: added interface ovn-k8s-mp0_0 on port 14
2022-11-18T21:01:14.773Z|00240|connmgr|INFO|br-ex<->unix#540: 3 flow_mods in the last 0 s (3 adds)
2022-11-18T21:01:23.053Z|00241|connmgr|INFO|br-int<->unix#533: 1815 flow_mods in the 1 s starting 10 s ago (1801 adds, 14 deletes)
2022-11-18T21:01:33.626Z|00242|bridge|INFO|bridge br-int: deleted interface ovn-k8s-mp0_0 on port 14
2022-11-18T21:01:33.811Z|00243|bridge|WARN|could not open network device ovn-k8s-mp0_0 (No such device)
2022-11-18T21:01:34.464Z|00244|bridge|WARN|could not open network device ovn-k8s-mp0_0 (No such device)
```
```
Back to ovnkube-node logs:
I1118 21:01:34.507104   51185 ovs.go:200] Exec(38): /usr/bin/ovs-vsctl --timeout=15 --if-exists get interface ovn-k8s-mp0_0 ofport
I1118 21:01:34.516700   51185 ovs.go:203] Exec(38): stdout: "-1\n"
I1118 21:01:34.516724   51185 ovs.go:204] Exec(38): stderr: ""
I1118 21:01:34.516734   51185 ovs.go:200] Exec(39): /usr/bin/ovs-ofctl --no-stats --no-names dump-flows br-int table=65,out_port=-1
I1118 21:01:34.524495   51185 ovs.go:203] Exec(39): stdout: ""
I1118 21:01:34.524513   51185 ovs.go:204] Exec(39): stderr: "2022-11-18T21:01:34Z|00001|ofp_port|WARN|Negative value -1 is not a valid port number.\novs-ofctl: -1 is not a valid OpenFlow port\n"
I1118 21:01:34.524523   51185 ovs.go:206] Exec(39): err: exit status 1
```

DmytroLinkin commented:
Hi @wizhaoredhat

I looked into ovnkube logs and it makes no sense to me.

First lines are related to managenemt netdevice interface:

```
I1118 21:01:33.316902   51185 management-port-dpu.go:199] Setup management port dpu host: ens1f0v0
```

From this line it is clear that netdevice link was found under it's original name (because it reports `link.Attrs().Name`). Which means that ovnkube did a lookup of existing management netdevice interface, but You didn't pasted preceding lines, so it is not clear did it found old interface or not.

```
I1118 21:01:33.419318   51185 ovs.go:200] Exec(5): /usr/bin/ovs-vsctl --timeout=15 set Open_vSwitch . external-ids:ovn-orig-mgmt-port-netdev-name=ens1f0v0
```

After VF was configured and renamed, its original name stored at OVS DB for later use which is done only for node mode "full".

Following lines are related to management representor interface:

```
I1118 21:01:33.584049   51185 ovs.go:200] Exec(6): /usr/bin/ovs-vsctl --timeout=15 list Interface ovn-k8s-mp0_0
I1118 21:01:33.595098   51185 ovs.go:200] Exec(7): /usr/bin/ovs-vsctl --timeout=15 get Interface ovn-k8s-mp0_0 type
I1118 21:01:33.605419   51185 management-port_linux.go:365] Found existing representor management port. Removing it
I1118 21:01:33.605430   51185 ovs.go:200] Exec(8): /usr/bin/ovs-vsctl --timeout=15 --if-exists get Interface ovn-k8s-mp0_0 external-ids:ovn-orig-mgmt-port-rep-name
I1118 21:01:33.615538   51185 ovs.go:200] Exec(9): /usr/bin/ovs-vsctl --timeout=15 --if-exists del-port br-int ovn-k8s-mp0_0
```

This lines suggests that representor link was found under original name, because in other case old port is *not looked up*. It was found, removed and renamed to original name `ens1f0_0` which is stored as external_ids. This is the 1st strange thing, because it is representor of the 1st VF, which is `ens1f0v0` from above. So I guess that representor somehow was renamed to ovn-k8s-mp0_0, but.. in such case this whole path of lookup old port shouldn't have been taken.

```
I1118 21:01:33.731455   51185 management-port-dpu.go:64] Create representor management port: ovn-k8s-mp0_0
```

Here we see that representor really was found not under original name, but, again, in such case ovnkube shouldn't even start looking for existing interface above. The only way I could imagine it happen, is that some other process in the parallel renamed representor.

```
I1118 21:01:33.795935   51185 ovs.go:200] Exec(10): /usr/bin/ovs-vsctl --timeout=15 -- --may-exist add-port br-int ovn-k8s-mp0_0 -- set interface ovn-k8s-mp0_0 external-ids:iface-id=k8s-worker-advnetlab48 external-ids:ovn-orig-mgmt-port-rep-name=ovn-k8s-mp0_0
I1118 21:01:33.812274   51185 ovs.go:203] Exec(10): stdout: ""
I1118 21:01:33.812290   51185 ovs.go:204] Exec(10): stderr: "ovs-vsctl: Error detected while setting up 'ovn-k8s-mp0_0': could not open network device ovn-k8s-mp0_0 (No such device).  See ovs-vswitchd log for details.\novs-vsctl: The default log directory is \"/var/log/openvswitch\".\n"
```

Here ovnkube tries to add representor to the bridge, but it disappeared? Was renamed? Does switchdev mode was disabled or SRIOV set to 0?

How come that representor was found under name `ovn-k8s-mp0_0` and ovnkube did old port lookup? For me it looks that You aren't using my code.

DmytroLinkin commented:
> Hi @wizhaoredhat
> 
> ...
> How come that representor was found under name `ovn-k8s-mp0_0` and ovnkube did old port lookup? For me it looks that You aren't using my code.

Ok, my bad. I found how it could happen.

wizhaoredhat commented:
Hi Dmytro, Unfortunately it is still failing:
```
I1121 16:11:06.332154  870257 management-port-dpu.go:174] Lookup netdevice link and existing management port
I1121 16:11:06.332290  870257 management-port-dpu.go:201] Setup netdevice management port: ovn-k8s-mp0
I1121 16:11:06.332304  870257 ovs.go:200] Exec(4): /usr/bin/ovs-vsctl --timeout=15 set Open_vSwitch . external-ids:ovn-orig-mgmt-port-netdev-name=ovn-k8s-mp0
I1121 16:11:06.340426  870257 ovs.go:203] Exec(4): stdout: ""
I1121 16:11:06.340440  870257 ovs.go:204] Exec(4): stderr: ""
I1121 16:11:06.384829  870257 management-port-dpu.go:46] Lookup representor link and existing management port
I1121 16:11:06.384945  870257 management-port-dpu.go:65] Setup representor management port: ovn-k8s-mp0_0
I1121 16:11:06.384975  870257 ovs.go:200] Exec(5): /usr/bin/ovs-vsctl --timeout=15 -- --may-exist add-port br-int ovn-k8s-mp0_0 -- set interface ovn-k8s-mp0_0 external-ids:iface-id=k8s-worker-advnetlab48 external-ids:ovn-orig-mgmt-port-rep-name=ovn-k8s-mp0_0
I1121 16:11:06.395539  870257 ovs.go:203] Exec(5): stdout: ""
I1121 16:11:06.395552  870257 ovs.go:204] Exec(5): stderr: ""
I1121 16:11:06.395577  870257 gateway_init.go:259] Initializing Gateway Functionality
...
I1121 16:11:06.559219  870257 gateway_shared_intf.go:1506] Shared Gateway Creation Complete
I1121 16:11:06.559349  870257 node.go:850] MTU (1500) of network interface br-ex is big enough to deal with Geneve header overhead (sum 1458). 
I1121 16:11:06.559360  870257 kube.go:97] Setting annotations map[k8s.ovn.org/l3-gateway-config:{"default":{"mode":"shared","interface-id":"br-ex_worker-advnetlab48","mac-address":"0c:42:a1:c6:ce:b8","ip-addresses":["192.168.111.33/24"],"ip-address":"192.168.111.33/24","next-hops":["192.168.111.1"],"next-hop":"192.168.111.1","node-port-enable":"true","vlan-id":"0"}} k8s.ovn.org/node-chassis-id:a79b6853-f662-4f99-846e-717631dcc491 k8s.ovn.org/node-mgmt-port-mac-address:0a:58:0a:83:00:02 k8s.ovn.org/node-primary-ifaddr:{"ipv4":"192.168.111.33/24"}] on node worker-advnetlab48
I1121 16:11:06.580413  870257 node.go:481] Waiting for gateway and management port readiness...
I1121 16:11:06.580464  870257 ovs.go:200] Exec(20): /usr/bin/ovs-vsctl --timeout=15 --if-exists get interface ovn-k8s-mp0_0 ofport
I1121 16:11:06.580487  870257 ovs.go:200] Exec(21): /usr/bin/ovs-appctl --timeout=15 -t /var/run/ovn/ovn-controller.869056.ctl connection-status
I1121 16:11:06.589283  870257 ovs.go:203] Exec(21): stdout: "connected\n"
I1121 16:11:06.589310  870257 ovs.go:204] Exec(21): stderr: ""
I1121 16:11:06.589317  870257 node.go:274] Node connection status = connected
I1121 16:11:06.589327  870257 ovs.go:200] Exec(22): /usr/bin/ovs-vsctl --timeout=15 -- br-exists br-int
I1121 16:11:06.590280  870257 ovs.go:203] Exec(20): stdout: "273\n"
I1121 16:11:06.590296  870257 ovs.go:204] Exec(20): stderr: ""
I1121 16:11:06.590304  870257 ovs.go:200] Exec(23): /usr/bin/ovs-ofctl --no-stats --no-names dump-flows br-int table=65,out_port=273
I1121 16:11:06.598884  870257 ovs.go:203] Exec(23): stdout: " cookie=0xdc92f1d5, table=65, priority=100,reg15=0x2,metadata=0xc actions=output:273\n"
I1121 16:11:06.598911  870257 ovs.go:204] Exec(23): stderr: ""
I1121 16:11:06.598919  870257 management-port.go:173] Management port ovn-k8s-mp0_0 is ready
I1121 16:11:06.599521  870257 ovs.go:203] Exec(22): stdout: ""
I1121 16:11:06.599535  870257 ovs.go:204] Exec(22): stderr: ""
I1121 16:11:06.599545  870257 ovs.go:200] Exec(24): /usr/bin/ovs-ofctl dump-aggregate br-int
I1121 16:11:06.608635  870257 ovs.go:203] Exec(24): stdout: "NXST_AGGREGATE reply (xid=0x4): packet_count=68612 byte_count=65187831 flow_count=1797\n"
I1121 16:11:06.608661  870257 ovs.go:204] Exec(24): stderr: ""
I1121 16:11:06.608673  870257 ovs.go:200] Exec(25): /usr/bin/ovs-vsctl --timeout=15 --if-exists get interface patch-br-ex_worker-advnetlab48-to-br-int ofport
I1121 16:11:06.617066  870257 ovs.go:203] Exec(25): stdout: "2\n"
I1121 16:11:06.617088  870257 ovs.go:204] Exec(25): stderr: ""
I1121 16:11:06.617094  870257 gateway.go:307] Gateway is ready
...
I1121 16:11:06.652046  870257 gateway_shared_intf.go:1655] Setting OVN Masquerade route with source: 192.168.111.33
I1121 16:11:06.652173  870257 ovs.go:200] Exec(28): /usr/sbin/ip route replace table 7 172.30.0.0/16 via 10.131.0.1 dev ovn-k8s-mp0
I1121 16:11:06.653746  870257 ovs.go:203] Exec(28): stdout: ""
I1121 16:11:06.653786  870257 ovs.go:204] Exec(28): stderr: ""
I1121 16:11:06.653803  870257 gateway_shared_intf.go:1356] Successfully added route into custom routing table: 7
...
I1121 16:11:06.853136  870257 node_upgrade.go:74] Detected cluster topology version 5 from ConfigMap openshift-ovn-kubernetes/control-plane-status
I1121 16:11:06.853170  870257 node.go:499] Current control-plane topology version is 5
I1121 16:11:06.853335  870257 ovs.go:200] Exec(34): /usr/bin/ovs-vsctl --timeout=15 --if-exists del-br br-ext
I1121 16:11:06.863579  870257 ovs.go:203] Exec(34): stdout: ""
I1121 16:11:06.863604  870257 ovs.go:204] Exec(34): stderr: ""
I1121 16:11:06.863617  870257 ovs.go:200] Exec(35): /usr/bin/ovs-vsctl --timeout=15 --if-exists del-port br-int int
I1121 16:11:06.868403  870257 ovs.go:203] Exec(33): stdout: ""
I1121 16:11:06.868429  870257 ovs.go:204] Exec(33): stderr: ""
I1121 16:11:06.873502  870257 ovs.go:203] Exec(35): stdout: ""
I1121 16:11:06.873522  870257 ovs.go:204] Exec(35): stderr: ""
F1121 16:11:06.873899  870257 ovnkube.go:133] unable to start health checking server: no mgmt ip
```
I see on the node that the ovn-k8s-mp0 interface is there but no IP address. The ovn-k8s-mp0_0 is there.:
```
20: ovn-k8s-mp0_0: <BROADCAST,MULTICAST> mtu 1400 qdisc mq state DOWN group default qlen 1000
    link/ether f6:a2:e9:2f:10:0a brd ff:ff:ff:ff:ff:ff
32: ovs-system: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 9e:3b:7a:4f:ae:92 brd ff:ff:ff:ff:ff:ff
33: genev_sys_6081: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 65000 qdisc noqueue master ovs-system state UNKNOWN group default qlen 1000
    link/ether 06:1e:4c:51:7f:ca brd ff:ff:ff:ff:ff:ff
    inet6 fe80::41e:4cff:fe51:7fca/64 scope link 
       valid_lft forever preferred_lft forever
34: br-int: <BROADCAST,MULTICAST> mtu 1400 qdisc noop state DOWN group default qlen 1000
    link/ether 8e:bf:ec:d5:a1:ae brd ff:ff:ff:ff:ff:ff
37: ovn-k8s-mp0: <BROADCAST,MULTICAST> mtu 1400 qdisc mq state DOWN group default qlen 1000
    link/ether 0a:58:0a:83:00:02 brd ff:ff:ff:ff:ff:ff
49: br-ex: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 0c:42:a1:c6:ce:b8 brd ff:ff:ff:ff:ff:ff
    inet 192.168.111.33/24 brd 192.168.111.255 scope global dynamic noprefixroute br-ex
       valid_lft 2179sec preferred_lft 2179sec
    inet 169.254.169.2/29 brd 169.254.169.7 scope global br-ex
       valid_lft forever preferred_lft forever
    inet6 fe80::e42:a1ff:fec6:ceb8/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
```

wizhaoredhat commented:
The pod keeps crashing eventually the error becomes:
```
I1121 20:35:18.349174 1317247 node.go:318] Detected support for port binding with external IDs
I1121 20:35:18.349258 1317247 ovs.go:200] Exec(4): /usr/bin/ovs-vsctl --timeout=15 list Interface ovn-k8s-mp0
I1121 20:35:18.358676 1317247 ovs.go:203] Exec(4): stdout: "_uuid               : eaeb19ea-99bb-4b95-9824-ac35f2ce2794\nadmin_state         : []\nbfd                 : {}\nbfd_status          : {}\ncfm_fault           : []\ncfm_fault_status    : []\ncfm_flap_count      : []\ncfm_health          : []\ncfm_mpid            : []\ncfm_remote_mpids    : []\ncfm_remote_opstate  : []\nduplex              : []\nerror               : \"could not add network device ovn-k8s-mp0 to ofproto (File exists)\"\nexternal_ids        : {iface-id=k8s-worker-advnetlab48}\nifindex             : []\ningress_policing_burst: 0\ningress_policing_kpkts_burst: 0\ningress_policing_kpkts_rate: 0\ningress_policing_rate: 0\nlacp_current        : []\nlink_resets         : []\nlink_speed          : []\nlink_state          : []\nlldp                : {}\nmac                 : []\nmac_in_use          : []\nmtu                 : []\nmtu_request         : 1400\nname                : ovn-k8s-mp0\nofport              : -1\nofport_request      : []\noptions             : {}\nother_config        : {}\nstatistics          : {}\nstatus              : {}\ntype                : internal\n"
I1121 20:35:18.358707 1317247 ovs.go:204] Exec(4): stderr: ""
I1121 20:35:18.358717 1317247 ovs.go:200] Exec(5): /usr/bin/ovs-vsctl --timeout=15 get Interface ovn-k8s-mp0 type
I1121 20:35:18.366795 1317247 ovs.go:203] Exec(5): stdout: "internal\n"
I1121 20:35:18.366815 1317247 ovs.go:204] Exec(5): stderr: ""
I1121 20:35:18.366824 1317247 ovs.go:200] Exec(6): /usr/bin/ovs-vsctl --timeout=15 list Interface ovn-k8s-mp0_0
I1121 20:35:18.375980 1317247 ovs.go:203] Exec(6): stdout: ""
I1121 20:35:18.375997 1317247 ovs.go:204] Exec(6): stderr: "ovs-vsctl: no row \"ovn-k8s-mp0_0\" in table Interface\n"
I1121 20:35:18.376002 1317247 ovs.go:206] Exec(6): err: exit status 1
I1121 20:35:18.376175 1317247 management-port_linux.go:411] Found existing management interface. Unconfiguring it
I1121 20:35:18.386844 1317247 ovs.go:200] Exec(7): /usr/bin/ovs-vsctl --timeout=15 --if-exists get Open_vSwitch . external-ids:ovn-orig-mgmt-port-netdev-name
I1121 20:35:18.396484 1317247 ovs.go:203] Exec(7): stdout: "ovn-k8s-mp0\n"
I1121 20:35:18.396497 1317247 ovs.go:204] Exec(7): stderr: ""
F1121 20:35:18.396569 1317247 ovnkube.go:133] failed to check existing management port: failed to rename ovn-k8s-mp0_0 link to ovn-k8s-mp0: file exists
```

wizhaoredhat commented on go-controller/pkg/node/management-port-dpu.go:
```diff
@@ -216,6 +232,13 @@ func (mp *managementPortNetdev) Create(nodeAnnotator kube.Annotator, waiter *sta
 		}
 	}
 
+	if config.OvnKubeNode.Mode != types.NodeModeDPUHost {
```
I think we need to check that mp.netdevName != k8sMgmtIntfName (a.k.a ovn-k8s-mp0) also mp.netdevName != k8sMgmtIntfName + "_0" (a.k.a ovn-k8s-mp0_0)

DmytroLinkin replied:
Additionally?

DmytroLinkin replied:
BTW, `mp.netdevName` can't be ovn-k8s-mp0, unless user specified this name as value for environment variable.

wizhaoredhat replied:
Yes additionally we need this check.

If you look at this log that I have:
```
I1121 16:11:06.332290  870257 management-port-dpu.go:201] Setup netdevice management port: ovn-k8s-mp0
I1121 16:11:06.332304  870257 ovs.go:200] Exec(4): /usr/bin/ovs-vsctl --timeout=15 set Open_vSwitch . **external-ids:ovn-orig-mgmt-port-netdev-name=ovn-k8s-mp0**
I1121 16:11:06.340426  870257 ovs.go:203] Exec(4): stdout: ""
I1121 16:11:06.340440  870257 ovs.go:204] Exec(4): stderr: ""
I1121 16:11:06.384829  870257 management-port-dpu.go:46] Lookup representor link and existing management port
I1121 16:11:06.384945  870257 management-port-dpu.go:65] Setup representor management port: ovn-k8s-mp0_0
I1121 16:11:06.384975  870257 ovs.go:200] Exec(5): /usr/bin/ovs-vsctl --timeout=15 -- --may-exist add-port br-int ovn-k8s-mp0_0 -- set interface ovn-k8s-mp0_0 external-ids:iface-id=k8s-worker-advnetlab48 **external-ids:ovn-orig-mgmt-port-rep-name=ovn-k8s-mp0_0**
I1121 16:11:06.395539  870257 ovs.go:203] Exec(5): stdout: ""
I1121 16:11:06.395552  870257 ovs.go:204] Exec(5): stderr: ""
```

This occurs because "mp.netdevName" can be "ovn-k8s-mp0". If you look at ```createNodeManagementPorts``` in ```node.go:327```. Let's assume ovn-kube-node previously ran with config.OvnKubeNode.MgmtPortNetdev = "ens1f0v0". This would rename "ens1f0v0" to "ovn-k8s-mp0" and "ens1f0_0" to "ovn-k8s-mp0_0". Next time ovn-kube-node node runs, config.OvnKubeNode.MgmtPortNetdev = "ens1f0v0" (Good). But then ```util.GetDeviceIDFromNetdevice(config.OvnKubeNode.MgmtPortNetdev)``` would fail since it was renamed. Then ```config.OvnKubeNode.MgmtPortNetdev = types.K8sMgmtIntfName``` is set on line 336. This overrides the value, so config.OvnKubeNode.MgmtPortNetdev = "ovn-k8s-mp0" . This would make "mp.netdevName" be "ovn-k8s-mp0" when calling "port.Create".

Same check needs to be done here:
```
management-port-dpu.go:92
// Plug management port representor to OVS.
if mp.repName != k8sMgmtIntfName {
	stdout, stderr, err := util.RunOVSVsctl(
		"--", "--may-exist", "add-port", "br-int", k8sMgmtIntfName,
		"--", "set", "interface", k8sMgmtIntfName,
		"external-ids:iface-id="+types.K8sPrefix+mp.nodeName,
		"external-ids:ovn-orig-mgmt-port-rep-name="+mp.repName)
} else {
	stdout, stderr, err := util.RunOVSVsctl(
		"--", "--may-exist", "add-port", "br-int", k8sMgmtIntfName,
		"--", "set", "interface", k8sMgmtIntfName,
		"external-ids:iface-id="+types.K8sPrefix+mp.nodeName)
}
```

DmytroLinkin replied:
I see. That make sense.

wizhaoredhat replied:
Thanks!, the latest changes look good.

DmytroLinkin commented:
> Hi Dmytro @DmytroLinkin , Unfortunately it is still failing:
> 
> ...

Hi. Even though error says about mgmt ip, it is totally different reason. Look onto this [PR](https://github.com/ovn-org/ovn-kubernetes/commit/7f0fb6a52eacd184335ba62ad9cc8a8f904519a2). It enables health check monitoring when OVN_EGRESSIP_HEALTHCHECK_PORT is not zero, but it is not zero by default. What You need to do is (besides of saying thanks to Flavio XD ) add OVN_EGRESSIP_HEALTHCHECK_PORT to env. variables with value "0".

DmytroLinkin commented:
> The pod keeps crashing eventually the error becomes:
> 
> ```
> I1121 20:35:18.349174 1317247 node.go:318] Detected support for port binding with external IDs
> I1121 20:35:18.349258 1317247 ovs.go:200] Exec(4): /usr/bin/ovs-vsctl --timeout=15 list Interface ovn-k8s-mp0
> I1121 20:35:18.358676 1317247 ovs.go:203] Exec(4): stdout: "_uuid               : eaeb19ea-99bb-4b95-9824-ac35f2ce2794\nadmin_state         : []\nbfd                 : {}\nbfd_status          : {}\ncfm_fault           : []\ncfm_fault_status    : []\ncfm_flap_count      : []\ncfm_health          : []\ncfm_mpid            : []\ncfm_remote_mpids    : []\ncfm_remote_opstate  : []\nduplex              : []\nerror               : \"could not add network device ovn-k8s-mp0 to ofproto (File exists)\"\nexternal_ids        : {iface-id=k8s-worker-advnetlab48}\nifindex             : []\ningress_policing_burst: 0\ningress_policing_kpkts_burst: 0\ningress_policing_kpkts_rate: 0\ningress_policing_rate: 0\nlacp_current        : []\nlink_resets         : []\nlink_speed          : []\nlink_state          : []\nlldp                : {}\nmac                 : []\nmac_in_use          : []\nmtu                 : []\nmtu_request         : 1400\nname                : ovn-k8s-mp0\nofport              : -1\nofport_request      : []\noptions             : {}\nother_config        : {}\nstatistics          : {}\nstatus              : {}\ntype                : internal\n"
> I1121 20:35:18.358707 1317247 ovs.go:204] Exec(4): stderr: ""
> I1121 20:35:18.358717 1317247 ovs.go:200] Exec(5): /usr/bin/ovs-vsctl --timeout=15 get Interface ovn-k8s-mp0 type
> I1121 20:35:18.366795 1317247 ovs.go:203] Exec(5): stdout: "internal\n"
> I1121 20:35:18.366815 1317247 ovs.go:204] Exec(5): stderr: ""
> I1121 20:35:18.366824 1317247 ovs.go:200] Exec(6): /usr/bin/ovs-vsctl --timeout=15 list Interface ovn-k8s-mp0_0
> I1121 20:35:18.375980 1317247 ovs.go:203] Exec(6): stdout: ""
> I1121 20:35:18.375997 1317247 ovs.go:204] Exec(6): stderr: "ovs-vsctl: no row \"ovn-k8s-mp0_0\" in table Interface\n"
> I1121 20:35:18.376002 1317247 ovs.go:206] Exec(6): err: exit status 1
> I1121 20:35:18.376175 1317247 management-port_linux.go:411] Found existing management interface. Unconfiguring it
> I1121 20:35:18.386844 1317247 ovs.go:200] Exec(7): /usr/bin/ovs-vsctl --timeout=15 --if-exists get Open_vSwitch . external-ids:ovn-orig-mgmt-port-netdev-name
> I1121 20:35:18.396484 1317247 ovs.go:203] Exec(7): stdout: "ovn-k8s-mp0\n"
> I1121 20:35:18.396497 1317247 ovs.go:204] Exec(7): stderr: ""
> F1121 20:35:18.396569 1317247 ovnkube.go:133] failed to check existing management port: failed to rename ovn-k8s-mp0_0 link to ovn-k8s-mp0: file exists
> ```

Because You switched back to internal OVS port, right?

wizhaoredhat commented:
> Hi. Even though error says about mgmt ip, it is totally different reason. Look onto this [PR](https://github.com/ovn-org/ovn-kubernetes/commit/7f0fb6a52eacd184335ba62ad9cc8a8f904519a2). It enables health check monitoring when OVN_EGRESSIP_HEALTHCHECK_PORT is not zero, but it is not zero by default. What You need to do is (besides of saying thanks to Flavio XD ) add OVN_EGRESSIP_HEALTHCHECK_PORT to env. variables with value "0".

Yes I disabled the Egress IP health check and it started work. Having the Egress IP health check  would cause OVN-K to crash when using the VF mgmt port. However it doesn't crash when using the internal OVS port. How can we get the VF mgmt port to work with Egress IP health check? (Yes I need to thank Flavio for this new problem =) )

> Because You switched back to internal OVS port, right?

No I did not switch back to internal OVS port. Ovn-kube-node kept crashing and eventually the log looked like this. I think we need the check before saving the original name in OvS. Otherwise it can get into this state if Ovn-kube-node restarts.

DmytroLinkin commented:
> > Hi. Even though error says about mgmt ip, it is totally different reason. Look onto this [PR](https://github.com/ovn-org/ovn-kubernetes/commit/7f0fb6a52eacd184335ba62ad9cc8a8f904519a2). It enables health check monitoring when OVN_EGRESSIP_HEALTHCHECK_PORT is not zero, but it is not zero by default. What You need to do is (besides of saying thanks to Flavio XD ) add OVN_EGRESSIP_HEALTHCHECK_PORT to env. variables with value "0".
> 
> Yes I disabled the Egress IP health check and it started work. Having the Egress IP health check would cause OVN-K to crash when using the VF mgmt port. However it doesn't crash when using the internal OVS port. How can we get the VF mgmt port to work with Egress IP health check? (Yes I need to thank Flavio for this new problem =) )

It is crashing not because of VF, but its representor, which eventually doesn't have IP. It happens not only for node mode full, but also for node mode DPU where only representor exists. So probably Flavio need to look onto DPU mode. For full mode I will think about solution, until that WA can be used.

> > Because You switched back to internal OVS port, right?
> 
> No I did not switch back to internal OVS port. Ovn-kube-node kept crashing and eventually the log looked like this. I think we need the check before saving the original name in OvS. Otherwise it can get into this state if Ovn-kube-node restarts.

I was confused by the logs, because some lines are missing there. I may guess, that on Your setup representor was removed from OVS bridge prior ovnkube-node start up, but wasn't renamed. By the logic, if interface is not on the bridge, then it is treated as VF netdevice, that is why ovnkube tried to do `ovn-k8s-mp0_0 -> ovn-k8s-mp0` renaming. How does representor got removed from the bridge?

wizhaoredhat commented:
/cc @flavio-fernandes

Hi Flavio, We are hitting some issues related to your changes to Egress IP HealthCheck https://github.com/ovn-org/ovn-kubernetes/commit/7f0fb6a52eacd184335ba62ad9cc8a8f904519a2 . In summary, in Full mode (with OvS HWOL) plus the use of a VF and VF rep as the Management Ports causes some issues because the VF rep mgmt port attached to the OvS bridge will not have an IP address. Rather the VF mgmt port will. In DPU mode (where the use of VF rep mgmt port is mandatory), there will be a case where the VF rep mgmt port will also have no IP address.

This causes the error "unable to start health checking server: no mgmt ip" to occur. I was wondering if you have any suggestions on solving this.

wizhaoredhat commented:
> > > Hi. Even though error says about mgmt ip, it is totally different reason. Look onto this [PR](https://github.com/ovn-org/ovn-kubernetes/commit/7f0fb6a52eacd184335ba62ad9cc8a8f904519a2). It enables health check monitoring when OVN_EGRESSIP_HEALTHCHECK_PORT is not zero, but it is not zero by default. What You need to do is (besides of saying thanks to Flavio XD ) add OVN_EGRESSIP_HEALTHCHECK_PORT to env. variables with value "0".
> > 
> > 
> > Yes I disabled the Egress IP health check and it started work. Having the Egress IP health check would cause OVN-K to crash when using the VF mgmt port. However it doesn't crash when using the internal OVS port. How can we get the VF mgmt port to work with Egress IP health check? (Yes I need to thank Flavio for this new problem =) )
> 
> It is crashing not because of VF, but its representor, which eventually doesn't have IP. It happens not only for node mode full, but also for node mode DPU where only representor exists. So probably Flavio need to look onto DPU mode. For full mode I will think about solution, until that WA can be used.

Got it. Will the Full mode solution be part of this PR? @flavio-fernandes Could we not error out if the mgmt ip is not present?

```
go-controller/pkg/node/node.go
@@ -660,7 +660,8 @@ func (n *OvnNode) startEgressIPHealthCheckingServer(wg *sync.WaitGroup, mgmtPort
                        return fmt.Errorf("failed start health checking server due to unsettled IPv6: %w", err)
                }
        } else {
-               return fmt.Errorf("unable to start health checking server: no mgmt ip")
+               klog.Infof("Unable to start Egress IP health checking server: no mgmt ip")
+               return nil
        }
```

> 
> > > Because You switched back to internal OVS port, right?
> > 
> > 
> > No I did not switch back to internal OVS port. Ovn-kube-node kept crashing and eventually the log looked like this. I think we need the check before saving the original name in OvS. Otherwise it can get into this state if Ovn-kube-node restarts.
> 
> I was confused by the logs, because some lines are missing there. I may guess, that on Your setup representor was removed from OVS bridge prior ovnkube-node start up, but wasn't renamed. By the logic, if interface is not on the bridge, then it is treated as VF netdevice, that is why ovnkube tried to do `ovn-k8s-mp0_0 -> ovn-k8s-mp0` renaming. How does representor got removed from the bridge?

Sorry I didn't want the log snippet to be too long. I am not sure what happend, but I guess with your previous patchset, if you were to kill the ovnkube-node container, you will see the problem. I tested your latest patchset and everything looks good!

DmytroLinkin commented:
> > > > Hi. Even though error says about mgmt ip, it is totally different reason. Look onto this [PR](https://github.com/ovn-org/ovn-kubernetes/commit/7f0fb6a52eacd184335ba62ad9cc8a8f904519a2). It enables health check monitoring when OVN_EGRESSIP_HEALTHCHECK_PORT is not zero, but it is not zero by default. What You need to do is (besides of saying thanks to Flavio XD ) add OVN_EGRESSIP_HEALTHCHECK_PORT to env. variables with value "0".
> > > ...
> 
> Got it. Will the Full mode solution be part of this PR?
> 

No, it is a fix. IMO, it should be a separate patch and include fix for DPU mode.

> > > > Because You switched back to internal OVS port, right?
> > > 
> > > 
> > > No I did not switch back to internal OVS port. Ovn-kube-node kept crashing and eventually the log looked like this. I think we need the check before saving the original name in OvS. Otherwise it can get into this state if Ovn-kube-node restarts.
> > 
> > 
> > I was confused by the logs, because some lines are missing there. I may guess, that on Your setup representor was removed from OVS bridge prior ovnkube-node start up, but wasn't renamed. By the logic, if interface is not on the bridge, then it is treated as VF netdevice, that is why ovnkube tried to do `ovn-k8s-mp0_0 -> ovn-k8s-mp0` renaming. How does representor got removed from the bridge?
> 
> Sorry I didn't want the log snippet to be too long. I am not sure what happend, but I guess with your previous patchset, if you were to kill the ovnkube-node container, you will see the problem. I tested your latest patchset and everything looks good!

From what I can say, representor might not be on the bridge because it was moved from it, or if the bridge itself was deleted. If it is a valid use case, then it probably should be handled correctly. But if You say that a latest patch fixed this...

wizhaoredhat commented:
> > > > > Hi. Even though error says about mgmt ip, it is totally different reason. Look onto this [PR](https://github.com/ovn-org/ovn-kubernetes/commit/7f0fb6a52eacd184335ba62ad9cc8a8f904519a2). It enables health check monitoring when OVN_EGRESSIP_HEALTHCHECK_PORT is not zero, but it is not zero by default. What You need to do is (besides of saying thanks to Flavio XD ) add OVN_EGRESSIP_HEALTHCHECK_PORT to env. variables with value "0".
> > > > > ...
> > 
> > 
> > Got it. Will the Full mode solution be part of this PR?
> 
> No, it is a fix. IMO, it should be a separate patch and include fix for DPU mode.

Agreed. Are you looking into it?

> 
> > > > > Because You switched back to internal OVS port, right?
> > > > 
> > > > 
> > > > No I did not switch back to internal OVS port. Ovn-kube-node kept crashing and eventually the log looked like this. I think we need the check before saving the original name in OvS. Otherwise it can get into this state if Ovn-kube-node restarts.
> > > 
> > > 
> > > I was confused by the logs, because some lines are missing there. I may guess, that on Your setup representor was removed from OVS bridge prior ovnkube-node start up, but wasn't renamed. By the logic, if interface is not on the bridge, then it is treated as VF netdevice, that is why ovnkube tried to do `ovn-k8s-mp0_0 -> ovn-k8s-mp0` renaming. How does representor got removed from the bridge?
> > 
> > 
> > Sorry I didn't want the log snippet to be too long. I am not sure what happend, but I guess with your previous patchset, if you were to kill the ovnkube-node container, you will see the problem. I tested your latest patchset and everything looks good!
> 
> From what I can say, representor might not be on the bridge because it was moved from it, or if the bridge itself was deleted. If it is a valid use case, then it probably should be handled correctly. But if You say that a latest patch fixed this...

I am not sure how it got into this state. But with the additional check in the latest patch set, I haven't seen it come up thus far.

flavio-fernandes commented:
> > > > Hi. Even though error says about mgmt ip, it is totally different reason. Look onto this [PR](https://github.com/ovn-org/ovn-kubernetes/commit/7f0fb6a52eacd184335ba62ad9cc8a8f904519a2). It enables health check monitoring when OVN_EGRESSIP_HEALTHCHECK_PORT is not zero, but it is not zero by default. What You need to do is (besides of saying thanks to Flavio XD ) add OVN_EGRESSIP_HEALTHCHECK_PORT to env. variables with value "0".
> > > 
> > > 
> > > Yes I disabled the Egress IP health check and it started work. Having the Egress IP health check would cause OVN-K to crash when using the VF mgmt port. However it doesn't crash when using the internal OVS port. How can we get the VF mgmt port to work with Egress IP health check? (Yes I need to thank Flavio for this new problem =) )
> > 
> > 
> > It is crashing not because of VF, but its representor, which eventually doesn't have IP. It happens not only for node mode full, but also for node mode DPU where only representor exists. So probably Flavio need to look onto DPU mode. For full mode I will think about solution, until that WA can be used.
> 
> Got it. Will the Full mode solution be part of this PR? @flavio-fernandes Could we not error out if the mgmt ip is not present?
> 

Yes. I will take care of that change via this Jira:  https://issues.redhat.com/browse/OCPBUGS-4098


> ```
> go-controller/pkg/node/node.go
> @@ -660,7 +660,8 @@ func (n *OvnNode) startEgressIPHealthCheckingServer(wg *sync.WaitGroup, mgmtPort
>                         return fmt.Errorf("failed start health checking server due to unsettled IPv6: %w", err)
>                 }
>         } else {
> -               return fmt.Errorf("unable to start health checking server: no mgmt ip")
> +               klog.Infof("Unable to start Egress IP health checking server: no mgmt ip")
> +               return nil
>         }
> ```
> 
> > > > Because You switched back to internal OVS port, right?
> > > 
> > > 
> > > No I did not switch back to internal OVS port. Ovn-kube-node kept crashing and eventually the log looked like this. I think we need the check before saving the original name in OvS. Otherwise it can get into this state if Ovn-kube-node restarts.
> > 
> > 
> > I was confused by the logs, because some lines are missing there. I may guess, that on Your setup representor was removed from OVS bridge prior ovnkube-node start up, but wasn't renamed. By the logic, if interface is not on the bridge, then it is treated as VF netdevice, that is why ovnkube tried to do `ovn-k8s-mp0_0 -> ovn-k8s-mp0` renaming. How does representor got removed from the bridge?
> 
> Sorry I didn't want the log snippet to be too long. I am not sure what happend, but I guess with your previous patchset, if you were to kill the ovnkube-node container, you will see the problem. I tested your latest patchset and everything looks good!

bn222 commented:
@wizhaoredhat @DmytroLinkin PTAL at https://github.com/ovn-org/ovn-kubernetes/pull/3292#discussion_r1034523743

bn222 commented:
yes, /lgtm

wizhaoredhat commented:
@DmytroLinkin Please remove the "Patches needs to be squashed" in the PR description and please rerun the failing ovn-ci test. Thanks!

DmytroLinkin commented:
/retest-failed

DmytroLinkin commented:
> @DmytroLinkin Please remove the "Patches needs to be squashed" in the PR description and please rerun the failing ovn-ci test. Thanks!

Done

wizhaoredhat commented:
@trozet Hi Tim, can you help us merge this PR?

trozet commented:
Yeah I will review this tmrw.

wizhaoredhat commented:
@trozet Just a friendly ping for reviewing this, thanks!

wizhaoredhat commented:
@trozet Just reaching to see if you had a chance to review this. Thanks!

trozet commented:
@wizhaoredhat sorry for the delay, looking now

trozet commented on go-controller/pkg/node/management-port_linux.go:
```diff
@@ -280,6 +293,138 @@ func createPlatformManagementPort(interfaceName string, localSubnets []*net.IPNe
+
+	if savedName == "" {
+		// rename to PortName + "-ddmmyyHHMMSS"
+		klog.Warningf("No saved management port netdevice name, falling back to generic name")
```
should you print the name of the port in the log message?

DmytroLinkin replied:
Yes

wizhaoredhat replied:
@DmytroLinkin I think it would be better to also print out the savedName in the logs. Just so we have all the info when debugging.

trozet commented on go-controller/pkg/node/management-port-dpu.go:
```diff
@@ -165,20 +175,34 @@ func newManagementPortNetdev(hostSubnets []*net.IPNet) ManagementPort {
+	} else if mp.netdevName != types.K8sMgmtIntfName {
+		if config.OvnKubeNode.Mode == types.NodeModeDPUHost {
+			// We do not expect OVS running here so just check existing interface
+			err = unconfigurePlatformMgmtInterface(mp.hostSubnets, types.K8sMgmtIntfName)
```
the comment says check existing interface, but then you configure the mgmt port. I'm confused.

DmytroLinkin replied:
I'll change the comment then

trozet commented on go-controller/pkg/node/management-port_linux.go:
```diff
@@ -280,6 +293,138 @@ func createPlatformManagementPort(interfaceName string, localSubnets []*net.IPNe
+
+	if savedName == "" {
+		// rename to PortName + "-ddmmyyHHMMSS"
+		klog.Warningf("No saved management port representor name, falling back to generic name")
```
same thing, should you log the name of the port

DmytroLinkin replied:
Yes

wizhaoredhat replied:
@DmytroLinkin Same as my above comment.

trozet reviewed (changes requested):
I'm not really sure what the names are supposed to be between the different modes and port types (netdev, internal, representor) so its hard to review this for me. Added some comments.

DmytroLinkin commented on go-controller/pkg/node/management-port.go:
```diff
@@ -70,6 +70,12 @@ func newManagementPort(nodeName string, hostSubnets []*net.IPNet) ManagementPort
 }
 
 func (mp *managementPort) Create(nodeAnnotator kube.Annotator, waiter *startupWaiter) (*managementPortConfig, error) {
+	for _, mgmtPortName := range []string{types.K8sMgmtIntfName, types.K8sMgmtIntfName + "_0"} {
```
> it would be nice if types held a constant for VF representor name. ...

I'd prefer to not, since in the future we suppose to add (to upstream) support for workload pods on the DPU. In approach we do have now there are 2 instances of ovnkube-node on the DPU (arm host) - full mode (using SF and its representor) and dpu mode (using VF representor). Within such approach representors must have different names. 

> The code is confusing around various places in the code, but it looks like the VF representor name should always be ovn-k8s-mp0_0. Is that right?
> ...
> Is the representor different names depending on OVNK full vs dpu mode?

Yes, it is. I kept original name for DPU mode.

trozet replied:
Is there a predictable naming scheme in each of these scenarios? Or will the names sometimes be random? Might be good to have some readme document explaining how the naming conventions work.

DmytroLinkin replied:
No. Names are predictable. In most cases it is `ovn-k8s-mp0`. Only in case of running full mode it is `ovn-k8s-mp0` and `ovn-k8s-mp0_0`. In case of 2 instances of ovnkube-node running on arm host, I mentioned above, we are planning to have mgmt port name in some way configurable. In approach we do have, we just pass mgmt port name as environment variable for full mode ovnkube-node. So all mgmt port interfaces will look like:

```
# ip -br link  | grep ovn
ovn-k8s-mp1_0    ...  # full mode SF representor
ovn-k8s-mp1      ...  # full mode SF
ovn-k8s-mp0      ...  # dpu mode VF representor
```

DmytroLinkin commented on go-controller/pkg/node/management-port.go:
```diff
@@ -109,6 +115,30 @@ func (mp *managementPort) Create(nodeAnnotator kube.Annotator, waiter *startupWa
 	return cfg, nil
 }
 
+func (mp *managementPort) checkExistingMgmtPort(mgmtPortName string) error {
```
So You want me to change name of the function, not just add comments?

trozet replied:
preferably

DmytroLinkin commented on go-controller/pkg/node/management-port.go:
```diff
@@ -109,6 +115,30 @@ func (mp *managementPort) Create(nodeAnnotator kube.Annotator, waiter *startupWa
 
+func (mp *managementPort) checkExistingMgmtPort(mgmtPortName string) error {
+	if _, stderr, err := util.RunOVSVsctl("list", "Interface", mgmtPortName); err != nil {
+		if !strings.Contains(stderr, "no row \""+mgmtPortName+"\" in table") {
```
> --no-headings --data bare --columns type find  interface  name=

So if command fails to find interface it will return empty string, right? That is a problem, since for VF representor this command also returns empty string

trozet replied:
OK then you can do:
[root@ovn-control-plane ~]# ovs-vsctl --no-headings --data bare --format csv --columns type,name find  interface  name=ovn-k8s-mp0
internal,ovn-k8s-mp0

DmytroLinkin commented:
/retest-failed

DmytroLinkin commented:
/retest-failed

wizhaoredhat commented:
@DmytroLinkin Tested, Looks good in my setup.
@trozet I think the changes that Dmytro made looks good. Could you PTAL?

wizhaoredhat commented:
/retest-failed
